File: Xenon_VSCode/Xenon/BossEye.ts
﻿import CBoss = require("Boss")
import CExplosion = require("Explosion")
import CBigExplosion = require("BigExplosion")
import ActorInfo = require("ActorInfo")
import enums = require("Enums");
import gsCControls = require("Controls");
import gsCTimer = require("Timer");
import CActor = require("Actor");
import CPlayGameState = require("PlayGameState");
import CExplode = require("Exploder");

class CBossEye extends CBoss {

    BOSSEYE_TOTAL: number = 6;
    m_eye_number: number;
    m_state: enums.BossEyeState = enums.BossEyeState.BOSSEYE_SHUT;

    constructor(playGameState: CPlayGameState) {
        super();
        this.m_playGameState = playGameState;
        this.m_eye_number = 0;
        this.m_state = enums.BossEyeState.BOSSEYE_SHUT;
        this.m_name = "Boss Eye";
    }

    //-------------------------------------------------------------

    public getActorInfo(): ActorInfo {
        this.m_actorInfo = this.m_scene.GetlistOfActors();
        return this.m_actorInfo.GetActorInfoListItem(enums.ActorInfoType.INFO_BOSSEYE);
    }

    //-------------------------------------------------------------

    public activate(): boolean {
        if (!this.isActive()) {
            this.m_timer.start();
        }
        return super.activate();
    }

    //-------------------------------------------------------------

    public kill(): void {
        var x: CExplosion = new CBigExplosion(this.m_playGameState);
        this.m_scene.addActor(x);
        x.setPosition(this.getPosition());
        x.activate();

        this.m_active_eyes--;
        super.kill();
    }

    //-------------------------------------------------------------

    public update(controls: gsCControls, gameTime: gsCTimer): boolean {
        if (this.m_state != enums.BossEyeState.BOSSEYE_OPEN) {
            this.m_is_hit = false;
        }
        if (this.m_shield == 0) {
            var explode = new CExplode(this);
            super.kill();
            return true;
        }

        this.m_sprite.setFrame(this.m_eye_number + this.m_state * this.BOSSEYE_TOTAL);
        return true;
    }

    //-------------------------------------------------------------

    public registerHit(energy: number, hitter: CActor): void {
        if (this.m_state == enums.BossEyeState.BOSSEYE_OPEN) {
            super.registerHit(energy, hitter);
        }
    }

    //-------------------------------------------------------------

    public setEyeNumber(eye_number: number): void {
        this.m_eye_number = eye_number;
    }

    //-------------------------------------------------------------

    public setState(state: enums.BossEyeState): void {
        this.m_state = state;
    }
}

export = CBossEye;

File: Xenon-Original_C++_Code/xenon/source/bosseye.cpp
//-------------------------------------------------------------
//
// Class:	CBossEye
//
// Author:	John M Phillips
//
// Started:	06/05/00
//
// Base:	CAlien
//
// Derived:	None
//
//-------------------------------------------------------------

#include "game.h"

//-------------------------------------------------------------

BossEyeState CBossEye::m_state = BOSSEYE_SHUT;

//-------------------------------------------------------------

CBossEye::CBossEye()
{
	m_eye_number = 0;
	m_state = BOSSEYE_SHUT;
}

//-------------------------------------------------------------

CBossEye::~CBossEye()
{
}

//-------------------------------------------------------------

bool CBossEye::activate()
{
	if (!isActive())
		m_timer.start();

	return CActor::activate();
}

//-------------------------------------------------------------

void CBossEye::kill()
{
	CExplosion *x = new CBigExplosion;
	m_scene->addActor(x);
	x->setPosition(getPosition());
	x->activate();

	m_active_eyes--;

	CActor::kill();
}

//-------------------------------------------------------------

void CBossEye::registerHit(int energy,CActor *hitter)
{
	if (m_state == BOSSEYE_OPEN)
		CActor::registerHit(energy,hitter);
}

//-------------------------------------------------------------

bool CBossEye::update(Controls *controls)
{
	if (m_state != BOSSEYE_OPEN)
		m_is_hit = false;

	if (m_shield == 0) {
		explode();
		kill();
		return true;
		}

	m_sprite.setFrame(m_eye_number + (int) m_state * BOSSEYE_TOTAL);

	return true;
}

//-------------------------------------------------------------

void CBossEye::setEyeNumber(int eye_number)
{
	m_eye_number = eye_number;
}

//-------------------------------------------------------------

void CBossEye::setState(BossEyeState state)
{
	m_state = state;
}

//-------------------------------------------------------------
